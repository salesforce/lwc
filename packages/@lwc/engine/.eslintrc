{
    "env": {
        "browser": true
    },

    "globals": {
        "process": true
    },

    "rules": {
        "lwc-internal/no-production-assert": "error"
    },

    "overrides": [
        {
            "files": [
                "**/__tests__/**"
            ],

            "rules": {
                "lwc-internal/no-production-assert": "off"
            }
        },
        {
            "files": [ "src/polyfills/**"],
            "rules": {
                "no-extend-native": "off"
            }
        }
    ]
}
