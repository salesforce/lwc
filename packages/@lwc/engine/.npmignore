coverage/
scripts/
test/
src/
tsconfig.json
global.d.ts
yarn.lock
.*
