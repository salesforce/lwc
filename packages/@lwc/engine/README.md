# LWC Engine

@lwc/engine is a private package that handles the runtime for LWC. Don't use this package directly; instead, use lwc.

This packgage includes:

-   Synthetic Shadow DOM
-   Diffing Algo
-   Accessibility Polyfill (AOM)
-   Native Shadow DOM support

### Dependencies

`@lwc/engine` does not include any external runtime dependencies. There are a few "dependencies" that we've manually imported, but it should be noted that this code has been heavilty modified and has little resemblance to the original public code.

-   Snabbdom
    -   Our diffing library for DOM nodes.
    -   src/3rdparty/snabbdom
-   Polymer
    -   Shadow DOM polyfills
    -   src/3rdparty/polymer
