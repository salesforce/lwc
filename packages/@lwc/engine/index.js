// eslint-disable-next-line no-console
console.log('@lwc/engine cannot be imported via cjs');
