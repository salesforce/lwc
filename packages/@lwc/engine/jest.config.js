/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const path = require('path');
const BASE_CONFIG = require('../../../scripts/jest/base.config');

module.exports = {
    ...BASE_CONFIG,

    displayName: 'lwc-engine',

    roots: ['<rootDir>/src'],

    // Customize setup for the engine tests.
    setupFilesAfterEnv: [path.resolve(__dirname, 'scripts/jest/setup-test.js')],
    moduleNameMapper: {
        'test-utils': path.resolve(__dirname, 'scripts/jest/test-utils.js'),
    },

    // Ignore jest custom setup scripts from the code coverage.
    coveragePathIgnorePatterns: [
        '<rootDir>/scripts/',
        '<rootDir>/src/polyfills',
        '<rootDir>/src/framework/wc.ts',
    ],

    // Override global threshold for the package. As we increase the test coverage we should increase
    // the threshold as well.
    coverageThreshold: {
        global: {
            functions: 65,
            lines: 75,
            branches: 70,
        },
    },
};
