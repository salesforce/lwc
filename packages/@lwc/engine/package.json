{
    "name": "@lwc/engine",
    "version": "1.0.2",
    "description": "LWC Engine",
    "main": "dist/engine.cjs.js",
    "module": "dist/engine.js",
    "typings": "types/index.d.ts",
    "license": "MIT",
    "scripts": {
        "clean": "rm -rf dist/ types/",
        "test": "jest",
        "build": "tsc --emitDeclarationOnly && rollup --config scripts/rollup/rollup.config.js"
    },
    "files": [
        "dist/",
        "types/"
    ],
    "dependencies": {
        "@lwc/features": "1.0.2",
        "@lwc/shared": "1.0.2",
        "observable-membrane": "0.26.1"
    },
    "devDependencies": {
        "@lwc/template-compiler": "1.0.2"
    },
    "lwc": {
        "modules": [
            {
                "name": "lwc",
                "path": "dist/engine.js"
            }
        ]
    },
    "publishConfig": {
        "access": "public"
    }
}
