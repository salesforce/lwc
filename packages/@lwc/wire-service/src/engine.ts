/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
// subtypes from lwc
export interface WireDef {
    params?: {
        [key: string]: string;
    };
    static?: {
        [key: string]: any;
    };
    adapter: any;
    method?: 1;
}
export interface ElementDef {
    // wire is optional on ElementDef but the lwc guarantees it before invoking wiring service hook
    wire: {
        [key: string]: WireDef;
    };
}
