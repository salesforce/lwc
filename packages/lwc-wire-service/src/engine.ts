// subtypes from lwc
export { Element, ComposableEvent } from 'lwc';
export interface WireDef {
    params?: {
        [key: string]: string;
    };
    static?: {
        [key: string]: any;
    };
    adapter: any;
    method?: 1;
}
export interface ElementDef {
    // wire is optional on ElementDef but the lwc guarantees it before invoking wiring service hook
    wire: {
        [key: string]: WireDef
    };
}
