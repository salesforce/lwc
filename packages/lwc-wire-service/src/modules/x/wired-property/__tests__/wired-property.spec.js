import WiredProperty from 'x-wired-property';
import { createElement } from 'engine';
import { registerMockWireAdapters } from 'x-wire-adapter-mock';

describe('wired-property', () => {
    const {
        test: mockTestAdapter
    } = registerMockWireAdapters('test');

    afterEach(() => {
        while (document.body.firstChild) {
            document.body.removeChild(document.body.firstChild);
        }
    });

    describe('snapshots', () => {
        it('default snapshot', () => {
            const element = createElement('x-wired-property', { is: WiredProperty });
            document.body.appendChild(element);
            expect(element).toMatchSnapshot();
        });

        it('should display data correctly', () => {
            const element = createElement('x-wired-property', { is: WiredProperty });
            document.body.appendChild(element);
            mockTestAdapter.next({
                Name: 'name'
            });

            return Promise.resolve().then(() => {
                expect(element).toMatchSnapshot();
            });
        });

        it('should display error correctly', () => {
            const element = createElement('x-wired-property', { is: WiredProperty });
            document.body.appendChild(element);
            mockTestAdapter.error('error message');

            return Promise.resolve().then(() => {
                expect(element).toMatchSnapshot();
            });
        });

        it('should display complete correctly', () => {
            const element = createElement('x-wired-property', { is: WiredProperty });
            document.body.appendChild(element);
            mockTestAdapter.complete();

            return Promise.resolve().then(() => {
                expect(element).toMatchSnapshot();
            });
        });
    });

    describe('component lifecycle hooks', () => {
        it('should get data when created', () => {
            const element = createElement('x-wired-property', { is: WiredProperty });
            mockTestAdapter.next({
                Name: 'name'
            });

            document.body.appendChild(element);

            expect(element.textContent.substring('Name: '.length, element.textContent.indexOf('Error'))).toBe('name');
        });

        it('should stop receiving data when disconnected', () => {
            const element = createElement('x-wired-property', { is: WiredProperty });
            document.body.appendChild(element);
            mockTestAdapter.next({
                Name: 'name'
            });
            document.body.removeChild(element);
            mockTestAdapter.next({
                Name: 'new_name'
            });

            expect(element.textContent.substring('Name: '.length, element.textContent.indexOf('Error'))).toBe('');
        });

        it('should receive data when reconnected', () => {
            const element = createElement('x-wired-property', { is: WiredProperty });
            document.body.appendChild(element);
            mockTestAdapter.next({
                Name: 'name'
            });
            document.body.removeChild(element);
            mockTestAdapter.next({
                Name: 'new_name'
            });
            document.body.appendChild(element);

            expect(element.textContent.substring('Name: '.length, element.textContent.indexOf('Error'))).toBe('new_name');
        });
    });
});
