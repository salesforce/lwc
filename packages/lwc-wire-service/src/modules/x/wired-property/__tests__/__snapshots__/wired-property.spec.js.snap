// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wired-property snapshots default snapshot 1`] = `
<x-wired-property>
  Name: 
  
  <br />
  Error: 
</x-wired-property>
`;

exports[`wired-property snapshots should display complete correctly 1`] = `
<x-wired-property>
  Name: 
  
  <br />
  Error: 
</x-wired-property>
`;

exports[`wired-property snapshots should display data correctly 1`] = `
<x-wired-property>
  Name: 
  name
  <br />
  Error: 
</x-wired-property>
`;

exports[`wired-property snapshots should display error correctly 1`] = `
<x-wired-property>
  Name: 
  
  <br />
  Error: 
  error message
</x-wired-property>
`;
